[Feature] - Implement Playlist Specification

Brandon-T opened the pull request:
<!-- *Thank you for submitting a pull request, your contributions are greatly appreciated!* -->

## Security Review
https://github.com/brave/security/issues/382

## Summary of Changes
- Adds ability to play from notification center
- Adds ability to play from lockscreen
- Adds ability to play in the background and control the previous/next/seek
- Adds picture in picture
- Fixes fullscreen for iPhone and iPad
- Adds whitelist of sites for which the toast shows up
- Allows long press on videos to add them to playlist
- Adds picture in picture mode
- Adds ability to download videos and audio to disk
- Adds airplay to allow streaming to other AirPlay compliant devices
- Allows skipping to 15 seconds back and forward.
- Adds support for live streams and HLS streams (.m3u8)
- Adds support for repeat playing a single media item, repeating all media items, or not repeating at all.
- All requests are Ephemeral (all that can be controlled anyway)
- Shows thumbnails on lockscreen/notification center.

- Does not parse VTT for v1 release, so no subtitles/captions.
- Does not support WebM or base64 streams.
- Does not support Carplay for v1 release.
- Supports only what mime-types Apple/AVPlayer allows.

<!-- Enter a ticket number for this PR, create a new one if it is not there yet. -->
This pull request fixes #3368 

## Submitter Checklist:

- [ ] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:
<!-- Any useful notes explaining how best to test and verify. -->


## Screenshots:
<!-- If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here. -->


## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue is assigned to a milestone (should happen at merge time).

Brandon-T committed:
Fixing UIButton's label not showing in the centre of the background image.

Brandon-T committed:
Fixing UIButton's label not showing in the centre of the background image.

soner-yuksel committed:
Adding localization and organization to PlayListViewController

soner-yuksel committed:
Adding localization and organization to PlayListViewController

soner-yuksel committed:
Button highlighting for for skip forward and backward

soner-yuksel committed:
Button highlighting for for skip forward and backward

Brandon-T committed:
Added more assets and changed WKWebView to BraveWebView

Brandon-T committed:
Added more assets and changed WKWebView to BraveWebView

soner-yuksel committed:
Adding cache for images in the thumbnail, response, memory, disk

soner-yuksel committed:
Adding cache for images in the thumbnail, response, memory, disk

soner-yuksel committed:
Media Player Thumb Progress Icon End time changed to counting backwards
Removing Filter By

soner-yuksel committed:
Media Player Thumb Progress Icon End time changed to counting backwards
Removing Filter By

Brandon-T committed:
Fixing live streaming videos

Brandon-T committed:
Fixing live streaming videos

soner-yuksel committed:
Adding new icons and fixing some comstraints on video player

soner-yuksel committed:
Adding new icons and fixing some comstraints on video player

Brandon-T committed:
Fixing huge delay when fetching thumbnails

Brandon-T committed:
Fixing huge delay when fetching thumbnails

Brandon-T committed:
Fixed videos sources for certain sites and fixed controls not tappable.

Brandon-T committed:
Fixed videos sources for certain sites and fixed controls not tappable.

soner-yuksel committed:
iPad Presentation for alert and video screen is changed

soner-yuksel committed:
Changes iPad for PlayList View

soner-yuksel committed:
Changes iPad for PlayList View

soner-yuksel committed:
Reverting PlayList View Change

soner-yuksel committed:
Reverting PlayList View Change

soner-yuksel committed:
Adding Pull Down Presentation for phone and use fullscreen for iPad

soner-yuksel committed:
Adding Pull Down Presentation for phone and use fullscreen for iPad

Brandon-T committed:
Added Playlists TouchHandler

Brandon-T committed:
Added picture in picture

Brandon-T committed:
Adding Playlist Asset cache downloader.

Brandon-T committed:
Adding Playlist Asset cache downloader.

Brandon-T committed:
Fixing playback of cached videos. Apple's API does NOT download MP4's or container types. It only handles HLS streams, so we download all other container types manually.

Brandon-T committed:
Fixing playback of cached videos. Apple's API does NOT download MP4's or container types. It only handles HLS streams, so we download all other container types manually.

Brandon-T committed:
Fixed picture in picture button action.

Brandon-T committed:
Fixed picture in picture button action.

soner-yuksel committed:
Removing Playlist Header

soner-yuksel committed:
Removing Playlist Header

soner-yuksel committed:
iPad Layout Changes and basic skeleton

Brandon-T committed:
Fixing Playlist feedback:
- AutoPlay
- Shows toast on add
- Shows toast on auto-detect.
- Partial implementation of moving items order.
- Shows download percentage and download status.
- Adds playback rate/speed button.

Brandon-T committed:
Fixing Playlist feedback:
- AutoPlay
- Shows toast on add
- Shows toast on auto-detect.
- Partial implementation of moving items order.
- Shows download percentage and download status.
- Adds playback rate/speed button.

soner-yuksel committed:
List and Video controllers using communication

Brandon-T committed:
Lots of fixes!
Added toasts.
Fixed expiration handling.

soner-yuksel committed:
Fixing MErge Changes

soner-yuksel committed:
Localization added for playlist items

Brandon-T committed:
Fix expired media.

Brandon-T committed:
Partially fixing iOS bugs in PIP based on radar: https://developer.apple.com/forums/thread/659329
iOS 14 has a bug where if a direct layer is used, the layer is restored on top of ALL UI! IE: It will have the highest z-order, so you can't pause or exit the video EVER! Without killing the app. However, if you used a layer that is a sub-layer of another layer, it works fine :S
PIP still doesn't work when going in the background due to 16x9 ratio. Will fix in next commit.

Brandon-T committed:
Added more playlist states to the toast, fixed a lot of bugs related to playing items and viewing their durations.

Brandon-T committed:
Stop displaying toast forever.

Brandon-T committed:
Fixing adding items with 0 duration which shouldn't be possible anyway. We absolutely must have meta-data for the video and source in order to add it.

Brandon-T committed:
Removed Indicator Icon.

Brandon-T committed:
Fix playlist toast by dismissing it when you navigate to a new page.

Brandon-T committed:
Fixing live streams and audio detection bugs. I think my JS needs work tbh..

Brandon-T committed:
Fixing PIP on iOS 14 where backgrounding sets the layer to nil as per Apple recommendation, but this doesn't work if the layer is shared which happens with AVPictureInPictureController! So if PIP is active, we need to NOT set the layer nil and hope for the best..

Brandon-T committed:
Fixed PIP Control-Center playback not immediately playing.

Brandon-T committed:
Fixed iPad PIP to match iPhone.

soner-yuksel committed:
Add haptic feedback when video is added to playlist

soner-yuksel committed:
On video list reduce the spacing between the title and detail label

soner-yuksel committed:
Add spacing between buttons and also ttop padding to video player buttons at top in iPad

soner-yuksel committed:
Rename expired-video popu

soner-yuksel committed:
Show a prompt before deleting a video with swipe action

Brandon-T committed:
Fixed HLS thumbnail generation.

Brandon-T committed:
Fixed iPad HLS thumbnail generation

Brandon-T committed:
Fixed seeking bug introduced by refactoring of `isPlaying` boolean to be determined based on player's TimeControlStatus.

soner-yuksel committed:
Fixing Swipe gesture Problem on TrackBar to reveal list

Brandon-T committed:
Fixed deleting assets and cached assets.
Fixed downloading percentages not showing.

Brandon-T committed:
Fixed re-ordering of items in playlist via long-press-gesture.

Brandon-T committed:
Allow 1.0, 1.5, 2.0x rates.

soner-yuksel committed:
iPad Overlay changes

Brandon-T committed:
Fixed iPad navigation icon.
Added other icons like cloud icon which will be used for download indicator.

soner-yuksel committed:
Adding Playlist settings for toast show option and side bar option

soner-yuksel committed:
Moving the Upper video controls under the navigation bar

soner-yuksel committed:
Changing close bar button icon

Brandon-T committed:
Added Playlists TouchHandler

Brandon-T committed:
Fixed Playback Rate bug when seeking.
Fixed percentage display to display a whole integer.
Fixed downloading progress for files and hls streams.
Refactored all of downloading into its own download manager.

soner-yuksel committed:
Reverting Close Icon Change

Brandon-T committed:
Fixed playlist toast showing more than once due to metadata load in JS.

Brandon-T committed:
Fixing WebM videos and invalid video mime-types.

soner-yuksel committed:
The image and duration fix for iPad list controller

Brandon-T committed:
Update FullScreen logic for the new UI.

Brandon-T committed:
Implementing Drag-Drop.

soner-yuksel committed:
Some UI feedback fixes

Brandon-T committed:
Finished Re-ordering of playlist items.

Brandon-T committed:
Fixed maximum width of toast as per design.

Brandon-T committed:
Added new icons. Fixed width of controls to be maximum of 350 as per design.

Brandon-T committed:
Fixed domain prefix of www as per design.

soner-yuksel committed:
Show Toast for a defined list of Domains

Brandon-T committed:
Refactroring and updating so that thumbnails show on lockscreen.

Brandon-T committed:
Fixed crash on mainThread and block data URI streaming for now.

Brandon-T committed:
Added "Repeat" feature and "Repeat Toggle" feature.

Brandon-T committed:
Fixed Track Grabber Size.

Brandon-T committed:
Fixed typos.

Brandon-T committed:
Fix repeating for iPad.

Brandon-T committed:
Revert ad fix!
Fixed iPad thumbnail sizing.

soner-yuksel committed:
Fixing final Video removes the play button problem

jamesmudgett committed:
change menu order

jamesmudgett committed:
use Done instead of X for playlist close button

jamesmudgett committed:
fullscreen portrait layout, use title for info bar instead of domain, fallback to domain if title is empty

Brandon-T committed:
Fixed icons resizing and fixed icons not showing on the info bar.
Reverted some changes in other commits.

Brandon-T committed:
Fixing observer bug when seeking a video on a wbesite like youtube and the playlist item keeps triggering the observer.

Brandon-T committed:
Fixing last played item so it plays the item you last played even if it did expire and reloaded properly later.
If the item is truly expired, it will NOT save as the last played item as it doesn't actually play.

Brandon-T committed:
Fixed next button not working.

Brandon-T committed:
Fixed function not exporting as @objc properly.

soner-yuksel committed:
Removed Navigation Title with video name and add text title Media Player

soner-yuksel committed:
Changing X button toolbar to Done iPad

Brandon-T committed:
Fixed viewDidLoad not being called in landscape. Absolutely no idea why it happens :S
It isn't documented anywhere..

soner-yuksel committed:
Adding width max and mi values to set split-view list

soner-yuksel committed:
Disabling the autoplay when player is loaded

soner-yuksel committed:
Settings for First Video Auto Play is working

soner-yuksel committed:
Auto Download option to video playlist

jamesmudgett committed:
fix playlist table inset issues and scroll bars

Brandon-T committed:
Fixed bug in overlay and allowed user to tap away the overlay.

Brandon-T committed:
Added showing download size.

Brandon-T committed:
Fixed relative time.

Brandon-T committed:
Fixed CI not supporting inline playback..

kylehickinson reviewed (changes requested):
First off, huge props, this PR is absolutely massive and this feature is amazing! 🔥 This is my first pass of code review as due to the size its really hard to do it all in one go.

The main issue I'm seeing is we cannot have 2 separate VC's for iPad vs iPhone. iPad can be displayed in compact width size class mode, which causes it to behave like an iPhone. In the attached screenshot, the sidebar button does nothing. We should be using a `UISplitViewController` as the base regardless of iPhone/iPad and adjusting things in the `viewWillTransition(to:with:)` and `traitCollectionDidChange` methods. If this cannot be solved in time, I suggest we put off a specialized iPad layout until later.

![image](https://user-images.githubusercontent.com/529104/110380194-c4460300-8025-11eb-80e2-afa5eddee706.png)

iccub reviewed (changes requested):
nitpick/code style review part 1

There's still a lot of review, left bigger classes for later, adding this partial review to avoid duplicate comments with Kyle's review

iccub reviewed (changes requested):
nitpick/code style review part 2

My general feeling is that we could add more comments documentation for video player parts,
and we can modernize some parts of the code, mostly key-value observation, and make database layer prettier, nothing specific, we can think about it later

Brandon-T committed:
More Feedback

Brandon-T committed:
Fixing exiting fullscreen for split screen iPad.

Brandon-T committed:
Fixing split screen navigation controller being deallocated.

Brandon-T committed:
Fix crash when no items are available. This was a fix for iPhone 8.

soner-yuksel committed:
Adding Delete Cache and Delete Video Data options to Privacy settings and Clear All  to Playlist Settings as options

soner-yuksel committed:
Playlist Manager deleteAllItems feedback

Brandon-T committed:
Added much needed error handling. Refactored if-statements into switch cases.
Fixed the navigation bar not showing properly in split screen on certain devices.
Fixed crash.
Added long press to skip to previous or next track.

Brandon-T committed:
Kill PIP when the Playlist is cleared.

Brandon-T committed:
Fixed navigation bar not showing Done button on some devices.

Brandon-T committed:
Fixed constraints being broken on different device layouts due to fixed button size being intrinsic in a stackView.

Brandon-T committed:
Fix favIcon stretching.

soner-yuksel committed:
Deleting final item switches player to state of blank

Brandon-T committed:
Added double tap to change aspect ratio.
Fixed scaling for safe-area layout when in full-screen.

Brandon-T committed:
Block blobs from loading in playlist.

Brandon-T committed:
Workaround for iPad blob API supported on actual iPad but not on simulator.

Brandon-T committed:
Added picture in picture

Brandon-T committed:
Added playlist setting to fix iPad web-compatibility with MediaSource API.

Brandon-T committed:
Restrict MediaSource extension API to iPad only.

Brandon-T committed:
Fixing ads showing up in playlist for youtube.

Brandon-T committed:
Removed unused files.

Brandon-T committed:
Fixing function declaration feedback.

Brandon-T committed:
Moved data models and used DataController.
Fixed issues with ads showing up all the time and duplicate messages of expiry shown to the user.
Refactored downloading to use the AVAssetDownloadTask instead of AVAssetAggregateDownloadTask

soner-yuksel committed:
Adding Auto Download to add Item to the list

Brandon-T committed:
Addressing Feedback.

Brandon-T committed:
Add to playlist’ button to share sheet activity

Brandon-T committed:
Fixing rebase after project file target screw up.

Brandon-T committed:
Update to Player V2 UI

Brandon-T committed:
Use item duration instead of date added on playlist cell.

Brandon-T committed:
Retrieve duration for 90% of videos and show "Live Stream" if a media item is live.

Brandon-T committed:
Fixing playlist toast not disappearing on navigation change.

Brandon-T committed:
Huge update to mime-types to support all audio and video types that the iOS system/OS/AVPlayer supports.

Brandon-T committed:
Added debug functions to get a list of extensions and mime-types from file information.

Brandon-T committed:
Added more robust fallback detector for media type.

Brandon-T committed:
Determine proper layout when navigation bar isn't shown.
Fixed pip for when navigation isn't there so that it retains the correct controller.
Hide navigation bar on non-compact layouts.

Brandon-T committed:
iPad Presentation for alert and video screen is changed

Brandon-T committed:
iPad Layout Changes and basic skeleton

Brandon-T committed:
Merge origin/feature/playlists-v2 into feature/playlists-v2

Brandon-T committed:
Fixing PIP on iPhone (weird that it worked on iPad).

Brandon-T committed:
Fixed Playlist iPad layout and PIP.

Brandon-T committed:
Added landscape support.

Brandon-T committed:
presentation style fullscreen, status bar styling fix, adjust insets on player control bars, remove safe area margins on landscape

Brandon-T committed:
Fixed Feedback

Brandon-T committed:
Updated for iOS 13.

Brandon-T committed:
Addressing feedback.

Brandon-T committed:
Addressed feedback.

iccub reviewed (changes requested):
left few more comments

Brandon-T committed:
Addressed Feedback.

Brandon-T committed:
Adding back MainFrame javascript files.

Brandon-T committed:
Fixing the regressions from removing navigation bar and moving the toast code.

Brandon-T committed:
Possible fix for crash with rate observer.

soner-yuksel committed:
Changing Localization with updated text and removing unnecessary text

soner-yuksel committed:
Fixing additional localization elements

Brandon-T committed:
Permanent fix for ads by getting rid of invisible webview.

Brandon-T committed:
Fixing titles showing wrong.

Brandon-T committed:
Added support for last played item.
Refactored playing code into one place.
Added support to skip next expired track if the playback happens automatically and isn't user initiated.
Fixing video player indefinitely time from Int64 overflow when unsigned in formatter.

soner-yuksel committed:
Adding Playback Start option to settings playlist

Brandon-T committed:
Playlist Settings Changes and Strings

soner-yuksel committed:
Fixing the localized text for Expired Video pop-ups

Brandon-T committed:
Sandboxing playlist and injecting it at page start.

Brandon-T committed:
Fixing adblocking in invisible webview. Broke SoundCloud :(

Brandon-T committed:
Removed logging.

Brandon-T committed:
Removing scripts from sandbox and fixing automatic detection for soundcloud and having a timeout for invisible webView.

Brandon-T committed:
Set the default value of playlists to auto-download: on.

Brandon-T committed:
Addressed security feedback.

jumde reviewed (approved):
lgtm

Brandon-T committed:
Addressed all Feedback.

Brandon-T committed:
Added prompt for available disk space usage.

Brandon-T committed:
Added settings description for offline mode.

Brandon-T committed:
Renamed showOverlays

Brandon-T committed:
Refactor downloading.

iccub reviewed (changes requested):
Few small nitpicks, 2 files left, i will finish tomorrow

Brandon-T committed:
Possibly silence a wrong linting

kylehickinson reviewed (changes requested):
Please ensure you go through your changes and replace all `leftAnchor` with `leadingAnchor` and `rightAnchor` to `trailingAnchor` (and in SnapKit `.left` → `.leading, `.right` → `.trailing`) to better support RTL langauges.

Brandon-T committed:
Fixed leading and trailing.

Brandon-T committed:
Removed unused core data model.

iccub reviewed (changes requested):
Few more nitpicks, no big changes needed, we are very close to approve this code

Brandon-T committed:
Fixed constraints.

Brandon-T committed:
Address feedback.

iccub reviewed (approved):
So happy to hit the approve button :D 
amazing amount of work, good job

kylehickinson reviewed (approved):
Still would like to see BVC dupes somehow consolidated/abstracted, but we can work on that post merge

Also: 🎉🎉🎉🎉🎉
